app/src/main/java/org/schabi/newpipe/player/ui/MainPlayerUi.java

```diff
@@ -0,0 +1,988 @@
+        binding.queueButton.setAlpha(showQueue ? 1.0f : 0.0f);
+        binding.segmentsButton.setVisibility(showSegment ? View.VISIBLE : View.GONE);
+        binding.segmentsButton.setAlpha(showSegment ? 1.0f : 0.0f);
+    }
+
+    @Override
+    public void showSystemUIPartially() {
+        if (isFullscreen) {
+            getParentActivity().ifPresent(activity -> {
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+                    activity.getWindow().setStatusBarColor(Color.TRANSPARENT);
+                    activity.getWindow().setNavigationBarColor(Color.TRANSPARENT);
+                }
+                final int visibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE
+                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
+                        | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION;
+                activity.getWindow().getDecorView().setSystemUiVisibility(visibility);
+                activity.getWindow().clearFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
+            });
+        }
+    }
+
+    @Override
+    public void hideSystemUIIfNeeded() {
+        player.getFragmentListener().ifPresent(PlayerServiceEventListener::hideSystemUiIfNeeded);
+    }
+
+    /**
+     * Calculate the maximum allowed height for the {@link R.id.endScreen}
+     * to prevent it from enlarging the player.
+     * <p>
+     * The calculating follows these rules:
+     * <ul>
+     * <li>
+     *     Show at least stream title and content creator on TVs and tablets
+     *     when in landscape (always the case for TVs) and not in fullscreen mode.
+     *     This requires to have at least <code>85dp</code> free space for {@link R.id.detail_root}
+     *     and additional space for the stream title text size
+     *     ({@link R.id.detail_title_root_layout}).
+     *     The text size is <code>15sp</code> on tablets and <code>16sp</code> on TVs,
+     *     see {@link R.id.titleTextView}.
+     * </li>
+     * <li>
+     *     Otherwise, the max thumbnail height is the screen height.
+     *     TODO investigate why this is done on popup player, too
+     * </li>
+     * </ul>
+     *
+     * @param bitmap the bitmap that needs to be resized to fit the end screen
+     * @return the maximum height for the end screen thumbnail
+     */
+    @Override
+    protected float calculateMaxEndScreenThumbnailHeight(@NonNull final Bitmap bitmap) {
+        final int screenHeight = context.getResources().getDisplayMetrics().heightPixels;
+
+        if (DeviceUtils.isTv(context) && !isFullscreen()) {
+            final int videoInfoHeight =
+                    DeviceUtils.dpToPx(85, context) + DeviceUtils.spToPx(16, context);
+            return Math.min(bitmap.getHeight(), screenHeight - videoInfoHeight);
+        } else if (DeviceUtils.isTablet(context) && isLandscape() && !isFullscreen()) {
+            final int videoInfoHeight =
+                    DeviceUtils.dpToPx(85, context) + DeviceUtils.spToPx(15, context);
```

Comment:
Where are the "magical" constants 85, 16 and 15 coming from?
Also the tv and table procedures are nearly identical, maybe we should abstract them?

Reply:
The explanation is in the method javadoc. I created three constants, but I will not proceed further into abstracting (nothing good comes to mind) and/or improving the meaning of those values in this PR.